ipa-core/src/bin/report_collector.rs

```diff
@@ -171,25 +203,150 @@ fn gen_inputs(
     Ok(())
 }
 
+/// Panics
+/// if (security_model, query_style) tuple is undefined
+fn get_query_type(
+    security_model: IpaSecurityModel,
+    query_style: &IpaQueryStyle,
+    ipa_query_config: IpaQueryConfig,
+) -> QueryType {
+    match (security_model, query_style) {
+        (IpaSecurityModel::SemiHonest, IpaQueryStyle::Oprf) => {
+            QueryType::OprfIpaRelaxedDpPadding(ipa_query_config)
+        }
+        (IpaSecurityModel::Malicious, IpaQueryStyle::Oprf) => {
+            panic!("OPRF for malicious is not implemented as yet")
+        }
+    }
+}
+
+fn write_ipa_output_file(
+    path: &PathBuf,
+    query_result: &IpaQueryResult,
+) -> Result<(), Box<dyn Error>> {
+    // it will be sad to lose the results if file already exists.
+    let path = if Path::is_file(path) {
+        let mut new_file_name = thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(5)
+            .map(char::from)
+            .collect::<String>();
+        let file_name = path.file_stem().ok_or("not a file")?;
+
+        new_file_name.insert(0, '-');
+        new_file_name.insert_str(0, &file_name.to_string_lossy());
+        tracing::warn!(
+            "{} file exists, renaming to {:?}",
+            path.display(),
+            new_file_name
+        );
+
+        // it will not be 100% accurate until file_prefix API is stabilized
+        Cow::Owned(
+            path.with_file_name(&new_file_name)
+                .with_extension(path.extension().unwrap_or("".as_ref())),
+        )
+    } else {
+        Cow::Borrowed(path)
+    };
+    let mut file = File::options()
+        .write(true)
+        .create_new(true)
+        .open(path.deref())
+        .map_err(|e| format!("Failed to create output file {}: {e}", path.display()))?;
+
+    write!(file, "{}", serde_json::to_string_pretty(query_result)?)?;
+    Ok(())
+}
+
 async fn ipa(
     args: &Args,
-    network: &NetworkConfig,
     security_model: IpaSecurityModel,
     ipa_query_config: IpaQueryConfig,
     helper_clients: &[MpcHelperClient; 3],
     query_style: IpaQueryStyle,
 ) -> Result<(), Box<dyn Error>> {
-    let input = InputSource::from(&args.input);
-    let query_type: QueryType;
-    match (security_model, &query_style) {
-        (IpaSecurityModel::SemiHonest, IpaQueryStyle::Oprf) => {
-            query_type = QueryType::OprfIpaRelaxedDpPadding(ipa_query_config);
+    let query_type = get_query_type(security_model, &query_style, ipa_query_config);
+
+    let enc1 = &args
+        .enc_input_file1
+        .as_ref()
```

Comment:
is there a better way to do this without `.as_ref()`? the compiler keeps saying that it's required for `File::open`, but it smells like there should be a cleaner way to do this.

Reply:
you don't need `&` but I don't think you can get away w/o using `as_ref` to borrow the content of an option. See the suggestion above that should help with getting rid of options alltogether